src/storage/src/source/resumption.rs

```diff
@@ -134,9 +115,13 @@ where
 
                 cap_set.downgrade(new_upper.elements());
                 upper = new_upper;
+                // If the downstream source has finished, we exit early.
```

Comment:
nit: maybe some more explanation for why we use this as a proxy for "downstream source has shut down". Maybe:
```
The resumption frontier is lower bounded by the involved shards (data shard, remap shard, etc.), so if it goes to empty we know that the source has finished writing and can shut down.
```